### PR TITLE
new generic function agglomerateByVariable

### PR DESCRIPTION
**Pull request by theotimeee**
Create a new generic function `agglomerateByVariable` to replace the methods `mergeCols` and `mergeRows`.
`agglomerateByVariable` takes parameter `MARGIN` to specify wether to run `mergeCols` or `mergeRows`

### CODE REVIEW

File: R/agglomerate.R [hunk 1]
```diff
@@ -214,28 +287,63 @@ setMethod("agglomerateByRank", signature = c(x = "SummarizedExperiment"),
...
         x
     }
 )
 
+#' @rdname agglomerate-methods
+#' @aliases agglomerateByVariable
+#' @export
+setMethod("agglomerateByVariable", signature = c(x = "SummarizedExperiment"),
+            function(x, MARGIN, f, archetype = 1L, ...){
+                FUN <- switch(MARGIN,
```
Comment (TuomasBorman) [inline R/agglomerate.R]: Use .check_MARGIN function to convert MARGIN to 1 or 2

Comment (TuomasBorman) [reply R/agglomerate.R]: 1 and 2 is general way in R to denote rows and cols.

File: R/deprecate.R [hunk 2]
```diff
@@ -38,3 +38,128 @@ setMethod("taxonomyTree", signature = c(x = "SummarizedExperiment"),
...
+#' @rdname deprecate
+#' @aliases mergeSamples
+#' @export
+setMethod("mergeCols", signature = c(x = "SummarizedExperiment"),
+            function(x, ...){
+                .Deprecated(msg = paste0("'mergeCols' is deprecated.",
+                                        "Use agglomerateByVariable with 
+                                        parameter MARGIN = 'cols' instead"))
+                agglomerateByVariable(x, MARGIN = "cols", ...)
+            }
```
Comment (TuomasBorman) [inline R/deprecate.R]: The indentation of this message is off. It shows in the console like this --> test all this deprecated functions
```

Warning message:
'mergeCols' is deprecated.Use agglomerateByVariable with 
                                        parameter MARGIN = 'cols' instead 
```

File: R/agglomerate.R [hunk 3]
```diff
@@ -263,11 +371,14 @@ setMethod("agglomerateByRank", signature = c(x = "SingleCellExperiment"),
...
-              x <- agglomerateByRank(x, ..., altexp = altexp, strip_altexp = strip_altexp)
-              x
-          }
+            function(x, ..., altexp = NULL, strip_altexp = TRUE){
+                .Deprecated(old="agglomerateByRank", new="mergeFeaturesByRank",
+                            "Now agglomerateByRank is deprecated. 
+                            Use mergeFeaturesByRank instead.")
+                x <- agglomerateByRank(x, ..., altexp = altexp, 
+                                        strip_altexp = strip_altexp)
+                x
```
Comment (TuomasBorman) [inline R/agglomerate.R]: Move these to deprecate.R

These functions will be removed. only agglomerateByRank, agglomerateByPrevalence and agglomerateByVariable will stay

**Comment by TuomasBorman** [comment]
Check those comments. I think this looks good

File: R/agglomerate.R [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
-#' \code{\link{taxonomyRanks}} can be used.
+#' \code{\link{taxonomyRanks}} can be used. 
+#' 
+#' \code{agglomerateByVariable} merges data on rows or columns of a 
+#' \code{SummarizedExperiment} as defined by a \code{factor} alongside the 
+#' chosen dimension. Metadata from the \code{rowData} or \code{colData} are 
+#' retained as defined by \code{archetype}.
+#' \code{\link[SummarizedExperiment:SummarizedExperiment-class]{assay}} are 
+#' agglomerated, i.e. summed up. If the assay contains values other than counts 
+#' or absolute values, this can lead to meaningless values being produced.
```
Comment (TuomasBorman) [inline R/agglomerate.R]: User wonders what is the difference between *Rank and *Variable.

*Rank: the data is agglomerated by taking into account the whole taxonomy table. --> use if you want to agglomerate based on taxonomy rank.
*Variable: the data is agglomerated based on certain variable. --> use if you want to agglomerate based on some other variable

Try to rephrase that

**Review by TuomasBorman — APPROVED** [review]

**Review by antagomir — APPROVED** [review]

**Comment by antagomir** [comment]
Seems good! Conflicts to be resolved

**Pull request merged**